File: Rosetta/Build League Table.R
# Build League Table.R
# Convert EPL Match data into league table

library(tidyverse) # (ggplot, tidyr, dplyr) - load data from CSV, tidy, transform and plot
library(readxl) # read from Excel

# Load the data from an Excel file
match_xlsx_filepath = "C:/Users/markw/OneDrive - Zomalex Ltd/Public/LBAG Online/Rosetta/EPL Data.xlsx"
match_full <-
  read_xlsx(path = match_xlsx_filepath, sheet = "EPL Match")

match <-
  match_full %>%
  select(
    Date,
    HomeTeam = `Home Team`,
    AwayTeam = `Away Team`,
    HomeGoals = `Full Time Home Goals`,
    AwayGoals = `Full Time Away Goals`
  )

# Step 1: Split and reshape
home_stats <-
  match %>%
  select(
    Date,
    Team = HomeTeam,
    OpposingTeam = AwayTeam,
    GF = HomeGoals,
    GA = AwayGoals
  )

away_stats <-
  match %>%
  select(
    Date,
    Team = AwayTeam,
    OpposingTeam = HomeTeam,
    GF = AwayGoals,
    GA = HomeGoals
  )

# Step 2: Append
team_stats <- home_stats %>%  union(away_stats)

# Step 3: Calculate
team_stats <-
  team_stats %>%
  mutate(
    Result = if_else(GF > GA, "W", if_else (GF == GA, "D", "L")),
    Points = if_else(Result == "W", 3, if_else(Result == "D", 1, 0)),
    Won = if_else(Result == "W", 1, 0),
    Drawn = if_else(Result == "D", 1, 0),
    Lost = if_else(Result == "L", 1, 0)
  )

# Step 4: Group By, Aggregate
league_table <-
  team_stats %>%
  group_by(Team) %>%
  summarise(
    Played = n(),
    Won = sum(Won),
    Drawn = sum(Drawn),
    Lost = sum(Lost),
    GF = sum(GF),
    GA = sum(GA),
    Points = sum(Points)
  ) %>%
  ungroup() %>%
  mutate(GD = GF - GA)  %>%  # Step 5: Calculate (again) GD = GF - GA
  arrange(-Points,-GD) # Step 6: Sort

File: Rosetta/build_league_table.py
import pandas as pd
import numpy as np

# Import the match results into a pandas dataframe
df_match = pd.read_csv("../EPL Data 20-21.csv")


result_col_list = ('Team', 'Opposing Team', 'GF', 'GA')

#  Each match generates tow results - one for the home teams and one for teh away team
home_cols_list = ['Home Team', 'Away Team', 'Full Time Home Goals', 'Full Time Away Goals']
df_home = df_match.loc[:, home_cols_list]
df_home.columns = result_col_list # ('Team', 'Opposing Team', 'GF', 'GA')

# The columns are ordered from the away teams perspective
# e.g. for the away team , Goals For (GF) is the number of away goals
away_cols_list = ['Away Team', 'Home Team', 'Full Time Away Goals', 'Full Time Home Goals']
df_away = df_match.loc[:, away_cols_list]
df_away.columns =  result_col_list # ('Team', 'Opposing Team', 'GF', 'GA')

# This contains all results - two for each match - one for the home team and one for the away team
df_all = df_home.append(df_away)

# Calculate whether the match was won, drawn or lost by the team
df_all['won'] = np.where(df_all['GF'] > df_all['GA'], 1 , 0)
df_all['drawn'] = np.where(df_all['GF'] == df_all['GA'], 1 , 0)
df_all['lost'] = np.where(df_all['GF'] < df_all['GA'], 1 , 0)

#  Group by Teams to get the league table and the total won / drawn / lost
df_league = df_all.groupby(['Team']).sum()

#  A  match will either result ina win / drw / loss so summing these (since either 1 or 0) will return matches played
df_league['played'] = df_league['won'] + df_league['drawn'] + df_league['lost']
#  Tams get 3 points for a win, 1 for a draw and no points for a loss
df_league['points'] = 3 * df_league['won'] + df_league['drawn']

#  The goal difference (GD) 
df_league['GD'] = df_league['GF'] - df_league['GA']

# Sort by Points (high to low) and then by GD (also high to low)
df_league  = df_league.sort_values(by = ['points', 'GD'], ascending = False)

# Now that the league tabkle is in the proper order , assign positions
df_league['position'] = np.arange(len(df_league)) + 1

#  Put the columns in the usual exapected order
df_league = df_league.loc[:, ['position', 'played', 'won', 'drawn', 'lost', 'GF', 'GA', 'GD', 'points']]

#  Write out to an Excel files
df_league.to_excel("league_table.xlsx")